File: src/humanoid_locomotion/humanoid_locomotion/facial_driver.py
# -*- coding: utf-8 -*-

import rclpy
from rclpy.node import Node
import serial
import yaml
import os
from std_msgs.msg import String
import time
import glob  
import termios

SERIAL_RECOVERABLE_ERRORS = (serial.SerialException, OSError, termios.error)

class FacialDriver(Node):
    def __init__(self):
        super().__init__('facial_driver_node')

        # 1. 参数配置
        self.declare_parameter('port', '/dev/ttyUSB0')
        self.declare_parameter('baudrate', 115200)
        self.declare_parameter('config_path', 'src/humanoid_locomotion/config/facial_gestures.yaml') 
        
        self.port = self.get_parameter('port').value
        self.baudrate = self.get_parameter('baudrate').value
        config_path = self.get_parameter('config_path').value
        
        # 2. 串口初始化
        self.ser = None
        self.connect_serial()


        # 3. 加载 YAML 动作库
        self.gestures = {}
        self.load_config(config_path)

        # 4. 订阅控制信号
        self.subscription = self.create_subscription(
            String,
            '/robot/facial_raw_cmd',
            self.callback,
            10
        )

        # 5. 中断控制变量
        self.is_interrupted = False
        self.current_gesture = "idle"

        # 6. 初始化：执行 idle 状态
        self.execute_gesture("idle")

    def load_config(self, path):
        """解析 YAML 配置文件"""
        try:
            if not os.path.isabs(path):
                path = os.path.join(os.getcwd(), path)

            if not os.path.exists(path):
                self.get_logger().error(f"❌ 配置文件不存在: {path}")
                return

            with open(path, 'r', encoding='utf-8') as f:
                data = yaml.safe_load(f)
                self.gestures = data.get('facial_gestures', {})
                self.get_logger().info(f"📚 成功加载表情库，共 {len(self.gestures)} 个动作")
        except Exception as e:
            self.get_logger().error(f"❌ 配置文件解析失败: {e}")

    def callback(self, msg):
        """处理收到的动作指令（支持中断）"""
        action_name = msg.data.strip()
        if not action_name:
            self.get_logger().warning("⚠️ 收到空指令，忽略")
            return

        self.get_logger().info(f"🛑 收到新表情指令: {action_name}，准备中断当前动作")
        self.is_interrupted = True
        time.sleep(0.5)  

        # 清空串口缓冲区，避免指令堆积
        self.clear_serial_buffers(log_success=True)

        self.is_interrupted = False
        self.current_gesture = action_name
        self.execute_gesture(action_name)

    def execute_gesture(self, name):
        """执行动作序列"""

        # 对所有复合/普通动作，先终止眼部循环；只有 eye_stop自己不需要前置停止
        if name in self.gestures and name not in ["eye_stop", "mouth_speak_stop"]:

            self.get_logger().info("🔍 先终止眼部循环和嘴部动作循环...")

            # 发送2次eye_stop指令，避免丢包
            if "eye_stop" in self.gestures:
                for _ in range(2):
                    for cmd in self.gestures["eye_stop"]:
                        if not self.send_raw(cmd):
                            return
                        time.sleep(0.1) #已修改原0.3
                self.get_logger().info("✅ 强制终止眼部循环（发送1次终止指令）")

            # 嘴部终止逻辑同步优化
            if "mouth_speak_stop" in self.gestures:
                for _ in range(1): #已修改原2
                    for cmd in self.gestures["mouth_speak_stop"]:
                        if not self.send_raw(cmd):
                            return
                        time.sleep(0.1) #已修改原0.3
                self.get_logger().info("✅ 强制终止嘴部循环（发送1次终止指令）")

            # 眼部强制复位（眼球居中+正常睁眼）
            eye_reset_cmds = ['$DGL:1!\n'] #已删除'$DGB:1!\n'
            for cmd in eye_reset_cmds:
                if not self.send_raw(cmd):
                    return
                time.sleep(0.1) #已修改原0.2
            self.get_logger().info("🔄 眼部强制复位（眼球居中+正常睁眼）")

        if name not in self.gestures:
            self.get_logger().warning(f"⚠️ 表情库中无 {name}，尝试作为原始指令发送")
            self.send_raw(name)
            return

        commands = self.gestures[name]
        self.get_logger().info(f"🎬 执行表情: {name}")

        self.clear_serial_buffers()

        for cmd in commands:
            if self.is_interrupted:
                self.get_logger().info(f"⚠️ 动作被中断，停止发送 {name} 的剩余指令")
                break

            # 解析自定义延迟标记（比如 delay:0.2）
            if cmd.startswith('delay:'):
                try:
                    delay_time = float(cmd.split(':')[1])
                    time.sleep(delay_time)
                    self.get_logger().info(f"⏳ 延迟 {delay_time} 秒（渐变过渡）")
                except:
                    self.get_logger().warning(f"⚠️ 无效延迟指令: {cmd}")
                continue

            # 原有逻辑：处理普通指令
            if name == "idle" and '!normal_start' in cmd:
                time.sleep(0.5)
                if not self.send_raw(cmd):
                    break
                time.sleep(0.8)
            else:
                if not self.send_raw(cmd):
                    break

                if name == "stop":
                    time.sleep(0.6)
                elif '!silent_start' in cmd or '!normal_start' in cmd:
                    time.sleep(0.8)
                else:
                    time.sleep(0.1)

    def connect_serial(self):
        """Open the facial serial port if it is not already open."""
        if self.ser and self.ser.is_open:
            return True

        try:
            self.ser = serial.Serial(self.port, self.baudrate, timeout=0.1)
            self.get_logger().info(f"✅ 已连接到仿生头模块: {self.port}, baudrate={self.baudrate}")
            time.sleep(2.0)
            return True
        except Exception as e:
            self.get_logger().error(f"❌ 头部串口连接失败: {e}")
            self.ser = None
            return False

    def close_serial(self):
        """Close and forget a broken serial handle."""
        if self.ser:
            try:
                if self.ser.is_open:
                    self.ser.close()
            except Exception:
                pass
        self.ser = None

    def clear_serial_buffers(self, log_success=False):
        """Best-effort serial buffer cleanup that must not crash the ROS node."""
        if not self.ser:
            return False

        try:
            if not self.ser.is_open:
                self.close_serial()
                return False
            self.ser.reset_input_buffer()
            self.ser.reset_output_buffer()
            if log_success:
                self.get_logger().info("🧹 清空串口缓冲区，避免指令残留")
            return True
        except SERIAL_RECOVERABLE_ERRORS as e:
            self.get_logger().error(f"❌ 清空串口缓冲区失败，将关闭串口并等待下次指令重连: {e}")
            self.close_serial()
            return False
        except Exception as e:
            self.get_logger().error(f"❌ 清空串口缓冲区异常，将关闭串口并等待下次指令重连: {e}")
            self.close_serial()
            return False

    def send_raw(self, cmd_str):
        """发送原始指令（自动补全\r\n，兼容硬件）"""
        if not cmd_str or cmd_str.strip() == "":
            self.get_logger().warning("⚠️ 空指令，忽略发送")
            return False

        if not self.connect_serial():
            self.get_logger().error("❌ 串口未连接，无法发送指令")
            return False

        try:
            send_cmd = cmd_str.strip() + "\r\n"
            self.ser.write(send_cmd.encode('utf-8'))
            self.get_logger().info(f"Serial Out: {send_cmd.strip()}")
            return True
        except SERIAL_RECOVERABLE_ERRORS as e:
            self.get_logger().error(f"❌ 串口发送失败，将关闭串口并等待下次指令重连: {e}")
            self.close_serial()
            return False
        except Exception as e:
            self.get_logger().error(f"❌ 指令发送异常: {e}")
            self.close_serial()
            return False

    def on_shutdown(self, send_sleep=False):
        """优雅关闭"""
        self.get_logger().info("🛑 开始优雅关闭...")
        if self.ser and self.ser.is_open:
            try:
                if send_sleep:
                    self.execute_gesture("sleeping")
                    time.sleep(1.0)
                self.close_serial()
                self.get_logger().info("🔌 串口已安全关闭")
            except Exception as e:
                self.get_logger().error(f"❌ 关闭串口失败: {e}")
        else:
            self.get_logger().info("🔌 串口未连接，无需关闭")


def main(args=None):
    rclpy.init(args=args)
    node = FacialDriver()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info("⚠️ 收到键盘中断信号（Ctrl+C）")
    except Exception as e:
        node.get_logger().error(f"❌ 节点运行异常: {e}")
    finally:
        node.on_shutdown()
        node.destroy_node()
        rclpy.shutdown()
        node.get_logger().info("✅ 节点已正常退出")


if __name__ == '__main__':
    main()

File: src/humanoid_expression_runtime/src/facial_driver.cpp
/*
 * 文件功能：C++ 表情串口驱动节点。
 *
 * 主要职责：
 * 1. 从参数文件读取表情动作库路径、串口设备、波特率和动作执行时序参数。
 * 2. 解析 humanoid_expression_runtime/config/facial_gestures.yaml 中的 facial_gestures 表情序列。
 * 3. 订阅 /robot/facial_raw_cmd，接收 APP 网关或调试工具发来的表情动作名。
 * 4. 按动作序列向仿生头串口控制板发送原始串口指令，并支持 delay:x 延时标记。
 * 5. 串口断开或发送失败时关闭句柄，下次动作自动尝试重连。
 * 6. 节点启动时执行 idle，退出时尝试执行 sleeping，与旧运行逻辑保持一致。
 *
 * 上游节点：
 * - humanoid_app_gateway_runtime/app_gateway_node：发布 /robot/facial_raw_cmd。
 * - facial_control_ui 或 ros2 topic pub：人工调试时也可直接发布 /robot/facial_raw_cmd。
 *
 * 下游硬件：
 * - 仿生头串口控制板，默认 /dev/serial/by-id/usb-1a86_USB_Serial-if00-port0。
 * - 如果 by-id 不存在，会按 fallback_ports 依次尝试 /dev/ttyUSB0、/dev/ttyUSB1 等候选串口。
 */

#include <fcntl.h>
#include <termios.h>
#include <unistd.h>

#include <cerrno>
#include <chrono>
#include <cstring>
#include <fstream>
#include <map>
#include <string>
#include <thread>
#include <vector>

#include <ament_index_cpp/get_package_share_directory.hpp>
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>
#include <yaml-cpp/yaml.h>

namespace
{

using namespace std::chrono_literals;

// 表情动作库：key 是动作名，例如 idle/talk；value 是串口指令和 delay 标记的有序列表。
using GestureLibrary = std::map<std::string, std::vector<std::string>>;

std::string trim_copy(const std::string & input)
{
  const auto begin = input.find_first_not_of(" \t\r\n");
  if (begin == std::string::npos) {
    return "";
  }
  const auto end = input.find_last_not_of(" \t\r\n");
  return input.substr(begin, end - begin + 1);
}

bool starts_with(const std::string & text, const std::string & prefix)
{
  return text.rfind(prefix, 0) == 0;
}

speed_t baudrate_to_termios(const int baudrate)
{
  switch (baudrate) {
    case 9600:
      return B9600;
    case 19200:
      return B19200;
    case 38400:
      return B38400;
    case 57600:
      return B57600;
    case 115200:
      return B115200;
    case 230400:
      return B230400;
    case 460800:
      return B460800;
    case 921600:
      return B921600;
    default:
      return B115200;
  }
}

}  // namespace

class FacialDriverCpp final : public rclcpp::Node
{
public:
  FacialDriverCpp()
  : Node("facial_driver_node")
  {
    declare_and_load_parameters();
    resolve_default_config_path();
    connect_serial();
    load_gesture_config();

    // 命令入口保持与旧 Python 节点一致：只接收纯动作名字符串，不接收 JSON。
    command_sub_ = create_subscription<std_msgs::msg::String>(
      command_topic_,
      rclcpp::QoS(10),
      [this](const std_msgs::msg::String::SharedPtr msg) {
        handle_command(msg->data);
      });

    if (!trim_copy(startup_gesture_).empty()) {
      current_gesture_ = startup_gesture_;
      execute_gesture(startup_gesture_);
    }
  }

  ~FacialDriverCpp() override
  {
    shutdown_driver();
  }

private:
  // ==================== 参数加载 ====================

  void declare_and_load_parameters()
  {
    declare_parameter<std::string>("config_path", "");
    declare_parameter<std::string>("command_topic", "/robot/facial_raw_cmd");
    declare_parameter<std::string>("port", "/dev/serial/by-id/usb-1a86_USB_Serial-if00-port0");
    declare_parameter<std::vector<std::string>>(
      "fallback_ports",
      std::vector<std::string>{"/dev/ttyUSB0", "/dev/ttyUSB1"});
    declare_parameter<int>("baudrate", 115200);
    declare_parameter<int>("serial_timeout_ms", 100);
    declare_parameter<double>("connect_settle_sec", 2.0);
    declare_parameter<double>("interrupt_wait_sec", 0.5);
    declare_parameter<std::string>("startup_gesture", "idle");
    declare_parameter<std::string>("shutdown_gesture", "sleeping");
    declare_parameter<bool>("pre_stop_eye_enable", true);
    declare_parameter<int>("pre_stop_eye_repeat", 2);
    declare_parameter<double>("pre_stop_eye_command_delay_sec", 0.1);
    declare_parameter<bool>("pre_stop_mouth_enable", true);
    declare_parameter<int>("pre_stop_mouth_repeat", 1);
    declare_parameter<double>("pre_stop_mouth_command_delay_sec", 0.1);
    declare_parameter<bool>("eye_reset_enable", true);
    declare_parameter<std::vector<std::string>>("eye_reset_commands", std::vector<std::string>{"$DGL:1!"});
    declare_parameter<double>("eye_reset_command_delay_sec", 0.1);
    declare_parameter<double>("idle_normal_start_pre_delay_sec", 0.5);
    declare_parameter<double>("idle_normal_start_post_delay_sec", 0.8);
    declare_parameter<double>("stop_command_delay_sec", 0.6);
    declare_parameter<double>("loop_start_command_delay_sec", 0.8);
    declare_parameter<double>("default_command_delay_sec", 0.1);
    declare_parameter<double>("shutdown_wait_sec", 1.0);
    declare_parameter<bool>("log_serial_output", true);

    get_parameter("config_path", config_path_);
    get_parameter("command_topic", command_topic_);
    get_parameter("port", port_);
    get_parameter("fallback_ports", fallback_ports_);
    get_parameter("baudrate", baudrate_);
    get_parameter("serial_timeout_ms", serial_timeout_ms_);
    get_parameter("connect_settle_sec", connect_settle_sec_);
    get_parameter("interrupt_wait_sec", interrupt_wait_sec_);
    get_parameter("startup_gesture", startup_gesture_);
    get_parameter("shutdown_gesture", shutdown_gesture_);
    get_parameter("pre_stop_eye_enable", pre_stop_eye_enable_);
    get_parameter("pre_stop_eye_repeat", pre_stop_eye_repeat_);
    get_parameter("pre_stop_eye_command_delay_sec", pre_stop_eye_command_delay_sec_);
    get_parameter("pre_stop_mouth_enable", pre_stop_mouth_enable_);
    get_parameter("pre_stop_mouth_repeat", pre_stop_mouth_repeat_);
    get_parameter("pre_stop_mouth_command_delay_sec", pre_stop_mouth_command_delay_sec_);
    get_parameter("eye_reset_enable", eye_reset_enable_);
    get_parameter("eye_reset_commands", eye_reset_commands_);
    get_parameter("eye_reset_command_delay_sec", eye_reset_command_delay_sec_);
    get_parameter("idle_normal_start_pre_delay_sec", idle_normal_start_pre_delay_sec_);
    get_parameter("idle_normal_start_post_delay_sec", idle_normal_start_post_delay_sec_);
    get_parameter("stop_command_delay_sec", stop_command_delay_sec_);
    get_parameter("loop_start_command_delay_sec", loop_start_command_delay_sec_);
    get_parameter("default_command_delay_sec", default_command_delay_sec_);
    get_parameter("shutdown_wait_sec", shutdown_wait_sec_);
    get_parameter("log_serial_output", log_serial_output_);
  }

  void resolve_default_config_path()
  {
    config_path_ = trim_copy(config_path_);
    if (!config_path_.empty()) {
      return;
    }

    try {
      config_path_ = ament_index_cpp::get_package_share_directory("humanoid_expression_runtime") +
        "/config/facial_gestures.yaml";
    } catch (const std::exception & e) {
      // 未 source install 或包未安装时，保留源码树兜底路径，方便开发期直接验证。
      (void)e;
      config_path_ = "src/humanoid_expression_runtime/config/facial_gestures.yaml";
    }
  }

  // ==================== YAML 表情库加载 ====================

  void load_gesture_config()
  {
    gestures_.clear();

    try {
      const YAML::Node root = YAML::LoadFile(config_path_);
      const YAML::Node facial_gestures = root["facial_gestures"];
      if (!facial_gestures || !facial_gestures.IsMap()) {
        RCLCPP_ERROR(get_logger(), "表情库格式错误，缺少 facial_gestures: %s", config_path_.c_str());
        return;
      }

      for (const auto & item : facial_gestures) {
        const std::string name = item.first.as<std::string>();
        const YAML::Node sequence = item.second;
        std::vector<std::string> commands;

        if (sequence && sequence.IsSequence()) {
          for (const auto & command : sequence) {
            commands.push_back(command.as<std::string>());
          }
        }
        gestures_[name] = commands;
      }

      RCLCPP_INFO(
        get_logger(),
        "成功加载表情库: path=%s, gestures=%zu",
        config_path_.c_str(),
        gestures_.size());
    } catch (const std::exception & e) {
      RCLCPP_ERROR(get_logger(), "表情库加载失败: path=%s, error=%s", config_path_.c_str(), e.what());
    }
  }

  // ==================== 串口生命周期 ====================

  bool connect_serial()
  {
    if (serial_fd_ >= 0) {
      return true;
    }

    const auto candidates = build_serial_candidates();
    std::string last_error;
    for (const auto & candidate : candidates) {
      if (open_and_configure_serial(candidate, last_error)) {
        active_port_ = candidate;
        RCLCPP_INFO(get_logger(), "已连接到仿生头模块: %s, baudrate=%d", active_port_.c_str(), baudrate_);
        sleep_seconds(connect_settle_sec_);
        return true;
      }
    }

    RCLCPP_ERROR(
      get_logger(),
      "头部串口连接失败，已尝试 %zu 个候选设备，最后错误: %s",
      candidates.size(),
      last_error.c_str());
    return false;
  }

  std::vector<std::string> build_serial_candidates() const
  {
    std::vector<std::string> candidates;
    auto append_unique = [&candidates](const std::string & port) {
      const std::string trimmed = trim_copy(port);
      if (trimmed.empty()) {
        return;
      }
      for (const auto & existing : candidates) {
        if (existing == trimmed) {
          return;
        }
      }
      candidates.push_back(trimmed);
    };

    append_unique(port_);
    for (const auto & fallback : fallback_ports_) {
      append_unique(fallback);
    }
    return candidates;
  }

  bool open_and_configure_serial(const std::string & candidate, std::string & last_error)
  {
    serial_fd_ = ::open(candidate.c_str(), O_RDWR | O_NOCTTY | O_SYNC);
    if (serial_fd_ < 0) {
      last_error = candidate + ": " + std::strerror(errno);
      RCLCPP_WARN(get_logger(), "候选串口打开失败: %s", last_error.c_str());
      return false;
    }

    termios tty {};
    if (::tcgetattr(serial_fd_, &tty) != 0) {
      last_error = candidate + ": 读取串口参数失败: " + std::strerror(errno);
      RCLCPP_WARN(get_logger(), "%s", last_error.c_str());
      close_serial();
      return false;
    }

    const speed_t speed = baudrate_to_termios(baudrate_);
    ::cfsetospeed(&tty, speed);
    ::cfsetispeed(&tty, speed);

    tty.c_cflag = (tty.c_cflag & ~CSIZE) | CS8;
    tty.c_iflag &= ~IGNBRK;
    tty.c_lflag = 0;
    tty.c_oflag = 0;
    tty.c_cc[VMIN] = 0;
    tty.c_cc[VTIME] = std::max(1, serial_timeout_ms_ / 100);
    tty.c_iflag &= ~(IXON | IXOFF | IXANY);
    tty.c_cflag |= (CLOCAL | CREAD);
    tty.c_cflag &= ~(PARENB | PARODD);
    tty.c_cflag &= ~CSTOPB;
    tty.c_cflag &= ~CRTSCTS;

    if (::tcsetattr(serial_fd_, TCSANOW, &tty) != 0) {
      last_error = candidate + ": 设置串口参数失败: " + std::strerror(errno);
      RCLCPP_WARN(get_logger(), "%s", last_error.c_str());
      close_serial();
      return false;
    }

    return true;
  }

  void close_serial()
  {
    if (serial_fd_ >= 0) {
      ::close(serial_fd_);
      serial_fd_ = -1;
    }
  }

  void flush_serial_buffers()
  {
    if (serial_fd_ < 0) {
      return;
    }

    if (::tcflush(serial_fd_, TCIOFLUSH) != 0) {
      RCLCPP_ERROR(get_logger(), "清空串口缓冲区失败，将尝试重连: %s", std::strerror(errno));
      close_serial();
    }
  }

  bool send_raw(const std::string & raw_command)
  {
    const std::string command = trim_copy(raw_command);
    if (command.empty()) {
      RCLCPP_WARN(get_logger(), "空指令，忽略发送");
      return false;
    }

    if (!connect_serial()) {
      RCLCPP_ERROR(get_logger(), "串口未连接，无法发送指令");
      return false;
    }

    const std::string payload = command + "\r\n";
    const char * data = payload.data();
    std::size_t remaining = payload.size();

    while (remaining > 0) {
      const ssize_t written = ::write(serial_fd_, data, remaining);
      if (written < 0) {
        if (errno == EINTR) {
          continue;
        }
        RCLCPP_ERROR(get_logger(), "串口发送失败，将关闭串口并等待下次指令重连: %s", std::strerror(errno));
        close_serial();
        return false;
      }
      data += written;
      remaining -= static_cast<std::size_t>(written);
    }

    if (log_serial_output_) {
      RCLCPP_INFO(get_logger(), "Serial Out: %s", command.c_str());
    }
    return true;
  }

  // ==================== ROS 命令入口 ====================

  void handle_command(const std::string & action)
  {
    const std::string action_name = trim_copy(action);
    if (action_name.empty()) {
      RCLCPP_WARN(get_logger(), "收到空指令，忽略");
      return;
    }

    RCLCPP_INFO(get_logger(), "收到新指令，中断当前动作: %s", action_name.c_str());
    is_interrupted_ = true;
    sleep_seconds(interrupt_wait_sec_);

    flush_serial_buffers();
    is_interrupted_ = false;
    current_gesture_ = action_name;
    execute_gesture(action_name);
  }

  // ==================== 动作执行 ====================

  void execute_gesture(const std::string & name)
  {
    const auto gesture_it = gestures_.find(name);
    const bool known_gesture = gesture_it != gestures_.end();

    // 对齐旧节点：普通表情执行前，先停眼部循环、停嘴部说话循环，再恢复睁眼。
    if (known_gesture && name != "eye_stop" && name != "mouth_speak_stop") {
      run_pre_stop_sequence();
    }

    if (!known_gesture) {
      RCLCPP_WARN(get_logger(), "表情库中无 %s，尝试作为原始指令发送", name.c_str());
      send_raw(name);
      return;
    }

    RCLCPP_INFO(get_logger(), "执行表情: %s", name.c_str());
    flush_serial_buffers();

    for (const auto & command : gesture_it->second) {
      if (is_interrupted_) {
        RCLCPP_INFO(get_logger(), "动作被中断，停止发送 %s 的剩余指令", name.c_str());
        break;
      }

      const std::string trimmed = trim_copy(command);
      if (starts_with(trimmed, "delay:")) {
        handle_delay_command(trimmed);
        continue;
      }

      if (name == "idle" && trimmed.find("!normal_start") != std::string::npos) {
        sleep_seconds(idle_normal_start_pre_delay_sec_);
        send_raw(trimmed);
        sleep_seconds(idle_normal_start_post_delay_sec_);
        continue;
      }

      send_raw(trimmed);

      if (name == "stop") {
        sleep_seconds(stop_command_delay_sec_);
      } else if (
        trimmed.find("!silent_start") != std::string::npos ||
        trimmed.find("!normal_start") != std::string::npos)
      {
        sleep_seconds(loop_start_command_delay_sec_);
      } else {
        sleep_seconds(default_command_delay_sec_);
      }
    }
  }

  void run_pre_stop_sequence()
  {
    if (pre_stop_eye_enable_) {
      run_named_sequence("eye_stop", pre_stop_eye_repeat_, pre_stop_eye_command_delay_sec_, "强制终止眼部循环");
    }

    if (pre_stop_mouth_enable_) {
      run_named_sequence(
        "mouth_speak_stop",
        pre_stop_mouth_repeat_,
        pre_stop_mouth_command_delay_sec_,
        "强制终止嘴部循环");
    }

    if (eye_reset_enable_) {
      for (const auto & command : eye_reset_commands_) {
        send_raw(command);
        sleep_seconds(eye_reset_command_delay_sec_);
      }
      RCLCPP_INFO(get_logger(), "眼部强制复位");
    }
  }

  void run_named_sequence(
    const std::string & name,
    const int repeat,
    const double command_delay_sec,
    const std::string & log_text)
  {
    const auto it = gestures_.find(name);
    if (it == gestures_.end()) {
      return;
    }

    const int safe_repeat = std::max(0, repeat);
    for (int round = 0; round < safe_repeat; ++round) {
      for (const auto & command : it->second) {
        send_raw(command);
        sleep_seconds(command_delay_sec);
      }
    }
    RCLCPP_INFO(get_logger(), "%s", log_text.c_str());
  }

  void handle_delay_command(const std::string & command)
  {
    try {
      const double delay_sec = std::stod(command.substr(std::string("delay:").size()));
      sleep_seconds(delay_sec);
      RCLCPP_INFO(get_logger(), "延迟 %.3f 秒", delay_sec);
    } catch (const std::exception &) {
      RCLCPP_WARN(get_logger(), "无效延迟指令: %s", command.c_str());
    }
  }

  void sleep_seconds(const double seconds)
  {
    if (seconds <= 0.0) {
      return;
    }
    std::this_thread::sleep_for(std::chrono::duration<double>(seconds));
  }

  // ==================== 关闭清理 ====================

  void shutdown_driver()
  {
    if (shutdown_done_) {
      return;
    }
    shutdown_done_ = true;

    RCLCPP_INFO(get_logger(), "开始优雅关闭表情驱动");
    if (serial_fd_ >= 0) {
      if (!trim_copy(shutdown_gesture_).empty()) {
        execute_gesture(shutdown_gesture_);
      }
      sleep_seconds(shutdown_wait_sec_);
      close_serial();
      RCLCPP_INFO(get_logger(), "串口已安全关闭");
    } else {
      RCLCPP_INFO(get_logger(), "串口未连接，无需关闭");
    }
  }

  // ==================== 成员变量 ====================

  std::string config_path_;
  std::string command_topic_;
  std::string port_;
  std::vector<std::string> fallback_ports_{"/dev/ttyUSB0", "/dev/ttyUSB1"};
  std::string active_port_;
  int baudrate_{115200};
  int serial_timeout_ms_{100};
  double connect_settle_sec_{2.0};
  double interrupt_wait_sec_{0.5};
  std::string startup_gesture_{"idle"};
  std::string shutdown_gesture_{"sleeping"};
  bool pre_stop_eye_enable_{true};
  int pre_stop_eye_repeat_{2};
  double pre_stop_eye_command_delay_sec_{0.1};
  bool pre_stop_mouth_enable_{true};
  int pre_stop_mouth_repeat_{1};
  double pre_stop_mouth_command_delay_sec_{0.1};
  bool eye_reset_enable_{true};
  std::vector<std::string> eye_reset_commands_{"$DGL:1!"};
  double eye_reset_command_delay_sec_{0.1};
  double idle_normal_start_pre_delay_sec_{0.5};
  double idle_normal_start_post_delay_sec_{0.8};
  double stop_command_delay_sec_{0.6};
  double loop_start_command_delay_sec_{0.8};
  double default_command_delay_sec_{0.1};
  double shutdown_wait_sec_{1.0};
  bool log_serial_output_{true};

  int serial_fd_{-1};
  bool is_interrupted_{false};
  bool shutdown_done_{false};
  std::string current_gesture_{"idle"};
  GestureLibrary gestures_;

  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr command_sub_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<FacialDriverCpp>();
  try {
    rclcpp::spin(node);
  } catch (const std::exception & e) {
    RCLCPP_ERROR(node->get_logger(), "节点运行异常: %s", e.what());
  }
  node.reset();
  rclcpp::shutdown();
  return 0;
}
